ci: extend and update the reviewer groups

IgorMinar opened the pull request:
Update the pullapprove config to require multiple reviews for sensitive groups in order
to force distribution of knowledge and improve the review quality.

IgorMinar committed:
ci: extend and update the reviewer groups

Update the pullapprove config to require multiple reviews for sensitive groups in order
to force distribution of knowledge and improve the review quality.

josephperrott commented on .pullapprove.yml:
```diff
@@ -352,7 +350,7 @@ groups:
         - AndrewKushnir
         - atscott
-        - kara
+        - ~kara  # do not request reviews from kara, but allow her to approve PRs
```
I am unsure of if pullapprove will properly parse a command after the content on the line.

I think we should have the comment on the above line, or make a more global comment explaining what the `~` does.

IgorMinar replied:
as confirmed below in the comments - this is supported

josephperrott reviewed (approved):
LGTM

IgorMinar reviewed (commented):
/pullapprove-test

<details>
```yml
####################################################################################
####################################################################################
#                               Angular Code Ownership                             #
####################################################################################
####################################################################################
#
#  Configuration of code ownership and review approvals for the angular/angular repo.
#
#  More info: https://docs.pullapprove.com/
#
# =========================================================
#  General rules / philosophy
# =========================================================
#
#  - We trust that people do the right thing and won't approve changes they don't feel confident reviewing
#  - We enforce that only approved PRs are merged ensuring that unreviewed code isn't accidentally merged
#  - We distribute approval rights as much as possible to help us scale better
#  - Groups have one or two global approvers groups as fallbacks:
#    - @angular/fw-global-approvers: for approving minor changes, large-scale refactorings, and emergency situations.
#    - @angular/fw-global-approvers-for-docs-only-changes: for approving minor documentation-only changes that don't require engineering review
#  - A small number of file groups have very limited number of reviewers because incorrect changes to the files they guard would have serious consequences (e.g. security, public api)
#
#  Configuration nuances:
#
#  - This configuration works in conjunction with the protected branch settings that require all changes to be made via pull requests with at least one approval.
#  - This approval can come from an appropriate codeowner, or any repo collaborator (person with write access) if the PR is authored by a codeowner.
#  - All groups whose pullapprove rules are matched will be required for overall approval.
#
#  NOTE:
#    In the case of emergency, the repo administrators which include the current angular caretaker
#    can bypass this reviewer approval requirement, this is expected as a last resort and to be
#    done exceedingly sparingly.

####################################################################################
#  GitHub usernames
####################################################################################
#  aikidave - Dave Shevitz
#  alan-agius4 - Alan Agius
#  alxhub - Alex Rickabaugh
#  AndrewKushnir - Andrew Kushnir
#  andrewseguin - Andrew Seguin
#  atscott - Andrew Scott
#  ayazhafiz - Ayaz Hafiz
#  clydin - Charles Lyding
#  crisbeto - Kristiyan Kostadinov
#  dennispbrown - Denny Brown
#  devversion - Paul Gschwendtner
#  dgp1130 - Doug Parker
#  filipesilva - Filipe Silva
#  gkalpak - Georgios Kalpakas
#  gregmagolan - Greg Magolan
#  IgorMinar - Igor Minar
#  jbogarthyde - Judy Bogart
#  jelbourn - Jeremy Elbourn
#  JiaLiPassion - Jia Li
#  JoostK - Joost Koehoorn
#  josephperrott - Joey Perrott
#  juleskremer - Jules Kremer
#  kapunahelewong - Kapunahele Wong
#  kara - Kara Erickson
#  kyliau - Keen Yee Liau
#  manughub - Manu Murthy
#  matsko - Matias Niemela
#  mgechev - Minko Gechev
#  mhevery - Miško Hevery
#  michaelprentice - Michael Prentice
#  mmalerba - Miles Malerba
#  petebacondarwin - Pete Bacon Darwin
#  pkozlowski-opensource - Pawel Kozlowski
#  robwormald - Rob Wormald
#  StephenFluin - Stephen Fluin


####################################################################################
#  Approval Groups
####################################################################################
# =========================================================
# @angular/framework-global-approvers
# =========================================================
#  Used for approving minor changes, large-scale refactorings, and in emergency situations.
#
# IgorMinar
# jelbourn
# josephperrott
# mhevery
#
# =========================================================
# @angular/framework-global-approvers-for-docs-only-changes
# =========================================================
#  Used for approving minor documentation-only changes that don't require engineering review.
#
# aikidave
# gkalpak
# kapunahelewong
# petebacondarwin


version: 3

# Meta field that goes unused by PullApprove to allow for defining aliases to be
# used throughout the config.
meta:
  1: &can-be-global-approved "\"global-approvers\" not in groups.approved"
  2: &can-be-global-docs-approved "\"global-docs-approvers\" not in groups.approved"

# turn on 'draft' support
# https://docs.pullapprove.com/config/github-api-version/
# https://developer.github.com/v3/previews/#draft-pull-requests
github_api_version: "shadow-cat-preview"

pullapprove_conditions:
  - condition: "'WIP' not in title"
    unmet_status: pending
    explanation: "Waiting to send reviews as PR is WIP"
  - condition: "'PR state: WIP' not in labels"
    unmet_status: pending
    explanation: "Waiting to send reviews as PR is WIP"
  - condition: "not draft"
    unmet_status: pending
    explanation: "Waiting to send reviews as PR is in draft"


groups:
  # =========================================================
  #  Global Approvers
  #
  # All reviews performed for global approvals require using
  # the `Reviewed-for:` specifier to set the approval
  # specificity as documented at:
  # https://docs.pullapprove.com/reviewed-for/
  # =========================================================
  global-approvers:
    type: optional
    reviewers:
      teams:
        - framework-global-approvers
    reviews:
      request: 0
      required: 1
      reviewed_for: required

  # =========================================================
  #  Global Approvers For Docs
  #
  # All reviews performed for global docs approvals require
  # using the `Reviewed-for:` specifier to set the approval
  # specificity as documented at:
  # https://docs.pullapprove.com/reviewed-for/
  # =========================================================
  global-docs-approvers:
    type: optional
    reviewers:
      teams:
        - framework-global-approvers-for-docs-only-changes
    reviews:
      request: 0
      required: 1
      reviewed_for: required

  # =========================================================
  #  Framework: Animations
  # =========================================================
  fw-animations:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'packages/animations/**',
          'packages/platform-browser/animations/**',
          'aio/content/guide/animations.md',
          'aio/content/examples/animations/**',
          'aio/content/images/guide/animations/**',
          'aio/content/guide/complex-animation-sequences.md',
          'aio/content/guide/reusable-animations.md',
          'aio/content/guide/route-animations.md',
          'aio/content/guide/transition-and-triggers.md'
          ])
    reviewers:
      users:
        - matsko


  # =========================================================
  #  Framework: Compiler
  # =========================================================
  fw-compiler:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files.exclude('packages/compiler-cli/ngcc/**'), [
          'packages/compiler/**',
          'packages/examples/compiler/**',
          'packages/compiler-cli/**',
          'aio/content/guide/angular-compiler-options.md',
          'aio/content/guide/aot-compiler.md',
          'aio/content/guide/aot-metadata-errors.md',
          'aio/content/guide/template-typecheck.md '
          ])
    reviewers:
      users:
        - alxhub
        - AndrewKushnir
        - JoostK


  # =========================================================
  #  Framework: Compiler / ngcc
  # =========================================================
  fw-ngcc:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - files.include('packages/compiler-cli/ngcc/**')
    reviewers:
      users:
        - alxhub
        - gkalpak
        - JoostK
        - petebacondarwin


  # =========================================================
  #  Framework: Migrations
  # =========================================================
  fw-migrations:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - files.include("packages/core/schematics/**")
    reviewers:
      users:
        - alxhub
        - crisbeto
        - devversion


  # =========================================================
  #  Framework: Core
  # =========================================================
  fw-core:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files.exclude("packages/core/schematics/**"), [
          'packages/core/**',
          'packages/examples/core/**',
          'packages/common/**',
          'packages/platform-browser/**',
          'packages/examples/platform-browser/**',
          'packages/platform-browser-dynamic/**',
          'packages/platform-webworker/**',
          'packages/platform-webworker-dynamic/**',
          'packages/examples/common/**',
          'packages/docs/**',
          'aio/content/guide/accessibility.md',
          'aio/content/examples/accessibility/**',
          'aio/content/guide/architecture-components.md',
          'aio/content/guide/architecture-modules.md',
          'aio/content/guide/architecture-next-steps.md',
          'aio/content/guide/architecture-services.md',
          'aio/content/guide/architecture.md',
          'aio/content/examples/architecture/**',
          'aio/content/images/guide/architecture/**',
          'aio/content/guide/attribute-directives.md',
          'aio/content/examples/attribute-directives/**',
          'aio/content/images/guide/attribute-directives/**',
          'aio/content/guide/bootstrapping.md',
          'aio/content/examples/bootstrapping/**',
          'aio/content/guide/cheatsheet.md',
          'aio/content/guide/component-interaction.md',
          'aio/content/examples/component-interaction/**',
          'aio/content/images/guide/component-interaction/**',
          'aio/content/guide/component-styles.md',
          'aio/content/examples/component-styles/**',
          'aio/content/guide/dependency-injection.md',
          'aio/content/examples/dependency-injection/**',
          'aio/content/images/guide/dependency-injection/**',
          'aio/content/guide/dependency-injection-in-action.md',
          'aio/content/examples/dependency-injection-in-action/**',
          'aio/content/images/guide/dependency-injection-in-action/**',
          'aio/content/guide/dependency-injection-navtree.md',
          'aio/content/guide/dependency-injection-providers.md',
          'aio/content/guide/displaying-data.md',
          'aio/content/examples/displaying-data/**',
          'aio/content/images/guide/displaying-data/**',
          'aio/content/guide/dynamic-component-loader.md',
          'aio/content/examples/dynamic-component-loader/**',
          'aio/content/images/guide/dynamic-component-loader/**',
          'aio/content/guide/entry-components.md',
          'aio/content/guide/feature-modules.md',
          'aio/content/examples/feature-modules/**',
          'aio/content/images/guide/feature-modules/**',
          'aio/content/guide/frequent-ngmodules.md',
          'aio/content/images/guide/frequent-ngmodules/**',
          'aio/content/guide/hierarchical-dependency-injection.md',
          'aio/content/examples/hierarchical-dependency-injection/**',
          'aio/content/examples/providers-viewproviders/**',
          'aio/content/examples/resolution-modifiers/**',
          'aio/content/guide/lazy-loading-ngmodules.md',
          'aio/content/examples/lazy-loading-ngmodules/**',
          'aio/content/images/guide/lazy-loading-ngmodules/**',
          'aio/content/guide/lifecycle-hooks.md',
          'aio/content/examples/lifecycle-hooks/**',
          'aio/content/images/guide/lifecycle-hooks/**',
          'aio/content/examples/ngcontainer/**',
          'aio/content/guide/ngmodules.md',
          'aio/content/examples/ngmodules/**',
          'aio/content/guide/ngmodule-api.md',
          'aio/content/guide/ngmodule-faq.md',
          'aio/content/guide/ngmodule-vs-jsmodule.md',
          'aio/content/guide/module-types.md',
          'aio/content/guide/template-syntax.md',
          'aio/content/examples/built-in-template-functions/**',
          'aio/content/examples/event-binding/**',
          'aio/content/examples/interpolation/**',
          'aio/content/examples/template-syntax/**',
          'aio/content/images/guide/template-syntax/**',
          'aio/content/examples/binding-syntax/**',
          'aio/content/examples/property-binding/**',
          'aio/content/examples/attribute-binding/**',
          'aio/content/examples/two-way-binding/**',
          'aio/content/examples/built-in-directives/**',
          'aio/content/images/guide/built-in-directives/**',
          'aio/content/examples/template-reference-variables/**',
          'aio/content/examples/inputs-outputs/**',
          'aio/content/images/guide/inputs-outputs/**',
          'aio/content/examples/template-expression-operators/**',
          'aio/content/guide/pipes.md',
          'aio/content/examples/pipes/**',
          'aio/content/images/guide/pipes/**',
          'aio/content/guide/providers.md',
          'aio/content/examples/providers/**',
          'aio/content/guide/singleton-services.md',
          'aio/content/guide/set-document-title.md',
          'aio/content/examples/set-document-title/**',
          'aio/content/images/guide/set-document-title/**',
          'aio/content/guide/sharing-ngmodules.md',
          'aio/content/guide/structural-directives.md',
          'aio/content/examples/structural-directives/**',
          'aio/content/images/guide/structural-directives/**',
          'aio/content/guide/user-input.md',
          'aio/content/examples/user-input/**',
          'aio/content/images/guide/user-input/**'
          ])
    reviewers:
      users:
        - alxhub
        - AndrewKushnir
        - atscott
        - ~kara  # do not request reviews from Kara, but allow her to approve PRs
        - mhevery
        - pkozlowski-opensource


  # =========================================================
  #  Framework: Http
  # =========================================================
  fw-http:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'packages/common/http/**',
          'packages/examples/http/**',
          'aio/content/guide/http.md',
          'aio/content/examples/http/**',
          'aio/content/images/guide/http/**'
          ])
    reviewers:
      users:
        - alxhub
        - IgorMinar


  # =========================================================
  #  Framework: Elements
  # =========================================================
  fw-elements:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'packages/elements/**',
          'aio/content/examples/elements/**',
          'aio/content/images/guide/elements/**',
          'aio/content/guide/elements.md'
          ])
    reviewers:
      users:
        - andrewseguin
        - gkalpak


  # =========================================================
  #  Framework: Forms
  # =========================================================
  fw-forms:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'packages/forms/**',
          'packages/examples/forms/**',
          'aio/content/guide/forms.md',
          'aio/content/examples/forms/**',
          'aio/content/images/guide/forms/**',
          'aio/content/guide/forms-overview.md',
          'aio/content/examples/forms-overview/**',
          'aio/content/images/guide/forms-overview/**',
          'aio/content/guide/form-validation.md',
          'aio/content/examples/form-validation/**',
          'aio/content/images/guide/form-validation/**',
          'aio/content/guide/dynamic-form.md',
          'aio/content/examples/dynamic-form/**',
          'aio/content/images/guide/dynamic-form/**',
          'aio/content/guide/reactive-forms.md',
          'aio/content/examples/reactive-forms/**',
          'aio/content/images/guide/reactive-forms/**'
          ])
    reviewers:
      users:
        - AndrewKushnir


  # =========================================================
  #  Framework: i18n
  # =========================================================
  fw-i18n:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'packages/core/src/i18n/**',
          'packages/core/src/render3/i18n.ts',
          'packages/core/src/render3/i18n.md',
          'packages/core/src/render3/interfaces/i18n.ts',
          'packages/common/locales/**',
          'packages/common/src/i18n/**',
          'packages/common/src/pipes/date_pipe.ts',
          'packages/common/src/pipes/i18n_plural_pipe.ts',
          'packages/common/src/pipes/i18n_select_pipe.ts',
          'packages/common/src/pipes/number_pipe.ts',
          'packages/compiler/src/i18n/**',
          'packages/compiler/src/render3/view/i18n/**',
          'packages/compiler-cli/src/extract_i18n.ts',
          'packages/localize/**',
          'aio/content/guide/i18n.md',
          'aio/content/examples/i18n/**'
          ])
    reviewers:
      users:
        - AndrewKushnir
        - mhevery
        - petebacondarwin


  # =========================================================
  #  Framework: Platform Server
  # =========================================================
  fw-platform-server:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'packages/platform-server/**',
          'aio/content/guide/universal.md',
          'aio/content/examples/universal/**'
          ])
    reviewers:
      users:
        - alxhub
        - kyliau


  # =========================================================
  #  Framework: Router
  # =========================================================
  fw-router:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'packages/router/**',
          'packages/examples/router/**',
          'aio/content/guide/router.md',
          'aio/content/guide/router-tutorial.md',
          'aio/content/examples/router-tutorial/**',
          'aio/content/examples/router/**',
          'aio/content/images/guide/router/**'
          ])
    reviewers:
      users:
        - atscott


  # =========================================================
  #  Framework: Service Worker
  # =========================================================
  fw-service-worker:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'packages/service-worker/**',
          'packages/examples/service-worker/**',
          'aio/content/guide/service-worker-getting-started.md',
          'aio/content/examples/service-worker-getting-started/**',
          'aio/content/guide/app-shell.md',
          'aio/content/guide/service-worker-communications.md',
          'aio/content/guide/service-worker-config.md',
          'aio/content/guide/service-worker-devops.md',
          'aio/content/guide/service-worker-intro.md',
          'aio/content/images/guide/service-worker/**'
          ])
    reviewers:
      users:
        - alxhub
        - gkalpak
        - IgorMinar


  # =========================================================
  #  Framework: Upgrade
  # =========================================================
  fw-upgrade:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'packages/upgrade/**',
          'packages/common/upgrade/**',
          'packages/examples/upgrade/**',
          'aio/content/guide/upgrade.md',
          'aio/content/examples/upgrade-lazy-load-ajs/**',
          'aio/content/examples/upgrade-module/**',
          'aio/content/images/guide/upgrade/**',
          'aio/content/examples/upgrade-phonecat-1-typescript/**',
          'aio/content/examples/upgrade-phonecat-2-hybrid/**',
          'aio/content/examples/upgrade-phonecat-3-final/**',
          'aio/content/guide/upgrade-performance.md',
          'aio/content/guide/upgrade-setup.md',
          'aio/content/guide/ajs-quick-reference.md',
          'aio/content/examples/ajs-quick-reference/**'
          ])
    reviewers:
      users:
        - gkalpak
        - petebacondarwin


  # =========================================================
  #  Framework: Testing
  # =========================================================
  fw-testing:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files.exclude('packages/compiler-cli/**'), [
          '**/testing/**',
          'aio/content/guide/testing.md',
          'aio/content/examples/testing/**',
          'aio/content/images/guide/testing/**'
          ])
    reviewers:
      users:
        - AndrewKushnir
        - IgorMinar
        - pkozlowski-opensource


  # =========================================================
  #  Framework: Benchmarks
  # =========================================================
  fw-benchmarks:
    conditions:
      - *can-be-global-approved
      - >
        contains_any_globs(files, [
          'modules/benchmarks/**'
          ])
    reviewers:
      users:
        - IgorMinar
        - pkozlowski-opensource


  # =========================================================
  #  Framework: Playground
  # =========================================================
  fw-playground:
    conditions:
      - *can-be-global-approved
      - >
        contains_any_globs(files, [
          'modules/playground/**'
          ])
    reviewers:
      users:
        - IgorMinar
        - jelbourn
        - pkozlowski-opensource


  # =========================================================
  #  Framework: Security
  # =========================================================
  fw-security:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'packages/core/src/sanitization/**',
          'packages/core/test/linker/security_integration_spec.ts',
          'packages/compiler/src/schema/**',
          'packages/platform-browser/src/security/**',
          'aio/content/guide/security.md',
          'aio/content/examples/security/**',
          'aio/content/images/guide/security/**'
          ])
    reviewers:
      users:
        - IgorMinar
        - mhevery
        - jelbourn
        - pkozlowski-opensource
    reviews:
      request: -1   # request reviews from everyone
      required: 2   # require at least 2 approvals
      reviewed_for: required


  # =========================================================
  #  Bazel
  # =========================================================
  bazel:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'packages/bazel/**',
          ])
    reviewers:
      users:
        - IgorMinar
        - josephperrott
        - kyliau


  # =========================================================
  #  Language Service
  # =========================================================
  language-service:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'packages/language-service/**',
          'aio/content/guide/language-service.md',
          'aio/content/images/guide/language-service/**'
          ])
    reviewers:
      users:
        - ayazhafiz
        - kyliau


  # =========================================================
  #  zone.js
  # =========================================================
  zone-js:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'packages/zone.js/**',
          'aio/content/guide/zone.md'
          ])
    reviewers:
      users:
        - JiaLiPassion
        - mhevery


  # =========================================================
  #  Benchpress
  # =========================================================
  benchpress:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'packages/benchpress/**'
          ])
    reviewers:
      users:
        - alxhub
        - josephperrott


  # =========================================================
  #  Integration Tests
  # =========================================================
  integration-tests:
    conditions:
      - *can-be-global-approved
      - >
        contains_any_globs(files, [
          'integration/**'
          ])
    reviewers:
      users:
        - IgorMinar
        - josephperrott
        - mhevery


  # =========================================================
  #  Docs: Gettings Started & Tutorial
  # =========================================================
  docs-getting-started-and-tutorial:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'aio/content/guide/setup-local.md',
          'aio/content/images/guide/setup-local/**',
          'aio/content/tutorial/**',
          'aio/content/images/guide/toh/**',
          'aio/content/examples/toh-pt0/**',
          'aio/content/examples/toh-pt1/**',
          'aio/content/examples/toh-pt2/**',
          'aio/content/examples/toh-pt3/**',
          'aio/content/examples/toh-pt4/**',
          'aio/content/examples/toh-pt5/**',
          'aio/content/examples/toh-pt6/**',
          'aio/content/examples/getting-started-v0/**',
          'aio/content/examples/getting-started/**',
          'aio/content/start/**',
          'aio/content/images/guide/start/**'
          ])
    reviewers:
      users:
        - aikidave
        - IgorMinar
        - StephenFluin


  # =========================================================
  #  Docs: Marketing
  # =========================================================
  docs-marketing:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'aio/content/marketing/**',
          'aio/content/images/bios/**',
          'aio/content/images/marketing/**',
          'aio/content/file-not-found.md',
          'aio/content/license.md',
          'aio/content/navigation.json'
          ])
    reviewers:
      users:
        - aikidave
        - IgorMinar
        - StephenFluin

  # =========================================================
  #  Docs: Observables
  # =========================================================
  docs-observables:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'aio/content/guide/observables.md',
          'aio/content/examples/observables/**',
          'aio/content/guide/comparing-observables.md',
          'aio/content/examples/comparing-observables/**',
          'aio/content/guide/observables-in-angular.md',
          'aio/content/examples/observables-in-angular/**',
          'aio/content/guide/practical-observable-usage.md',
          'aio/content/examples/practical-observable-usage/**',
          'aio/content/guide/rx-library.md',
          'aio/content/examples/rx-library/**'
          ])
    reviewers:
      users:
        - alxhub


  # =========================================================
  #  Docs: Packaging, Tooling, Releasing
  # =========================================================
  docs-packaging-and-releasing:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'docs/PUBLIC_API.md',
          'docs/RELEASE_SCHEDULE.md',
          'aio/content/guide/npm-packages.md',
          'aio/content/guide/browser-support.md',
          'aio/content/guide/releases.md',
          'aio/content/guide/updating.md',
          'aio/content/guide/deprecations.md',
          'aio/content/guide/migration-renderer.md',
          'aio/content/guide/migration-undecorated-classes.md',
          'aio/content/guide/migration-dynamic-flag.md',
          'aio/content/guide/migration-injectable.md',
          'aio/content/guide/migration-localize.md',
          'aio/content/guide/migration-module-with-providers.md',
          'aio/content/guide/static-query-migration.md',
          'aio/content/guide/updating-to-version-9.md',
          'aio/content/guide/ivy-compatibility.md',
          'aio/content/guide/ivy-compatibility-examples.md'
          ])
    reviewers:
      users:
        - IgorMinar
        - jelbourn


  # =========================================================
  #  Docs: CLI
  # =========================================================
  docs-cli:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'aio/content/cli/**',
          'aio/content/guide/typescript-configuration.md',
          'aio/content/examples/setup/**',
          'aio/content/guide/build.md',
          'aio/content/images/guide/build/**',
          'aio/content/guide/cli-builder.md',
          'aio/content/examples/cli-builder/**',
          'aio/content/guide/deployment.md',
          'aio/content/images/guide/deployment/**',
          'aio/content/guide/file-structure.md',
          'aio/content/guide/ivy.md',
          'aio/content/guide/web-worker.md',
          'aio/content/guide/workspace-config.md',
          ])
    reviewers:
      users:
        - clydin
        - IgorMinar
        - mgechev


  # =========================================================
  #  Docs: CLI Libraries
  # =========================================================
  docs-libraries:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'aio/content/guide/creating-libraries.md',
          'aio/content/guide/libraries.md',
          'aio/content/guide/using-libraries.md'
          ])
    reviewers:
      users:
        - alan-agius4
        - IgorMinar
        - mgechev


  # =========================================================
  #  Docs: Schematics
  # =========================================================
  docs-schematics:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'aio/content/guide/schematics.md',
          'aio/content/guide/schematics-authoring.md',
          'aio/content/guide/schematics-for-libraries.md',
          'aio/content/images/guide/schematics/**',
          'aio/content/examples/schematics-for-libraries/**'
          ])
    reviewers:
      users:
        - alan-agius4
        - IgorMinar
        - mgechev


  # =========================================================
  #  Docs-infra
  # =========================================================
  docs-infra:
    conditions:
      - *can-be-global-approved
      - *can-be-global-docs-approved
      - >
        contains_any_globs(files, [
          'aio/*',
          'aio/aio-builds-setup/**',
          'aio/content/cli-src/**',
          'aio/content/examples/*',
          'aio/scripts/**',
          'aio/src/**',
          'aio/tests/**',
          'aio/tools/**',
          'aio/content/guide/docs-style-guide.md',
          'aio/content/examples/docs-style-guide/**',
          'aio/content/images/guide/docs-style-guide/**',
          'aio/content/guide/visual-studio-2015.md'
          ])
    reviewers:
      users:
        - gkalpak
        - IgorMinar
        - petebacondarwin


  # =========================================================
  #  Dev-infra
  # =========================================================
  dev-infra:
    conditions:
      - *can-be-global-approved
      - >
        contains_any_globs(files.exclude("CHANGELOG.md"), [
          '*',
          '.circleci/**',
          '.devcontainer/**',
          '.github/**',
          '.ng-dev/**',
          '.vscode/**',
          '.yarn/**',
          'dev-infra/**',
          'docs/BAZEL.md',
          'docs/CARETAKER.md',
          'docs/COMMITTER.md',
          'docs/DEBUG.md',
          'docs/DEBUG_COMPONENTS_REPO_IVY.md',
          'docs/DEVELOPER.md',
          'docs/GITHUB_PROCESS.md',
          'docs/PR_REVIEW.md',
          'docs/SAVED_REPLIES.md',
          'docs/TOOLS.md',
          'docs/TRIAGE_AND_LABELS.md',
          'goldens/*',
          'modules/*',
          'packages/*',
          'packages/examples/test-utils/**',
          'packages/private/**',
          'packages/examples/*',
          'scripts/**',
          'third_party/**',
          'tools/build/**',
          'tools/circular_dependency_test/**',
          'tools/contributing-stats/**',
          'tools/gulp-tasks/**',
          'tools/ngcontainer/**',
          'tools/npm/**',
          'tools/npm_integration_test/**',
          'tools/rxjs/**',
          'tools/saucelabs/**',
          'tools/size-tracking/**',
          'tools/source-map-test/**',
          'tools/symbol-extractor/**',
          'tools/testing/**',
          'tools/ts-api-guardian/**',
          'tools/tslint/**',
          'tools/utils/**',
          'tools/yarn/**',
          'tools/*',
          '**/*.bzl',
          '**/*.bazel'
          ])
    reviewers:
      users:
        - devversion
        - filipesilva
        - gkalpak
        - IgorMinar
        - josephperrott


  # =========================================================
  #  Public API
  # =========================================================
  public-api:
    conditions:
      - *can-be-global-approved
      - >
        contains_any_globs(files, [
          'goldens/public-api/**',
          'CHANGELOG.md',
          'docs/NAMING.md',
          'aio/content/guide/glossary.md',
          'aio/content/guide/styleguide.md',
          'aio/content/examples/styleguide/**',
          'aio/content/images/guide/styleguide/*'
          ])
    reviewers:
      users:
        - alxhub
        - IgorMinar
        - jelbourn
        - pkozlowski-opensource
    reviews:
      request: -1   # request reviews from everyone
      required: 3   # require at least 3 approvals
      reviewed_for: required


  # ================================================
  #  Size tracking
  # ================================================
  size-tracking:
    conditions:
      - *can-be-global-approved
      - >
        contains_any_globs(files, [
          'goldens/size-tracking/**'
          ])
    reviewers:
      users:
        - alxhub
        - IgorMinar
        - jelbourn
        - pkozlowski-opensource
    reviews:
      request: -1   # request reviews from everyone
      required: 2   # require at least 2 approvals
      reviewed_for: required


  # ================================================
  #  Circular dependencies
  # ================================================
  circular-dependencies:
    conditions:
      - *can-be-global-approved
      - >
        contains_any_globs(files, [
          'goldens/circular-deps/packages.json'
        ])
    reviewers:
      users:
        - IgorMinar
        - jelbourn
        - josephperrott
        - pkozlowski-opensource


####################################################################################
#  Special Cases
####################################################################################

  # =========================================================
  #  Code Ownership
  # =========================================================
  code-ownership:
    conditions:
      - *can-be-global-approved
      - >
        contains_any_globs(files, [
          '.pullapprove.yml'
          ])
    reviewers:
      users:
        - IgorMinar


  # ====================================================
  #  Catch all for if no groups match the code change
  # ====================================================
  fallback:
    conditions:
      - *can-be-global-approved
      # Groups which are found to have matching conditions are `active`
      # according to PullApprove. If no groups are matched and considered
      # active, we still want to have a review occur.
      - len(groups.active) == 0
    reviewers:
      users:
        - IgorMinar
```
</details>

IgorMinar commented:
@josephperrott pullapprove does support inline comments:

Parsed config from [the test output above](https://app.pullapprove.com/report/?url=https%3A//storage.googleapis.com/pullapprove-storage/reports/77e19c35-14fe-4189-98de-e805486d4b16.json%3FExpires%3D1596580229%26GoogleAccessId%3Dpullapprove-reporter%2540pullapprove.iam.gserviceaccount.com%26Signature%3DV33bNy9mPuJkNKtIHVVpdeYUtqBjJmzCtoywhxaXDmsGOEex6wAqgVEnI6a%252BsAPIzTqhVs3jFj%252Bd4i2XV8SzVELHqBjlXxJ201eJXTHlcZeKi4wT5sOnsiOC5NoJjTvSfzvn29yJCg40FTkB93lF2z4%252B4P2WmqFQeRPaDHTb2lYHAGivZ9vt9YKexbCDvPl2Ac%252FZajUUoS7YzxsvhTho8ruXGssE9ztkZrDs2x2sT1HvSAmVtM7Prmn5wC%252FMSxif03ghyBi3f%252FgH5Qy0uhHjib3lJkmajwliBaVUqi8roI0UNfsPL5YEgpo4ilBCPnkIUHCe43%252Fb2fY5A4iGFDI4Hg%253D%253D&fingerprint=e7b0a0f826d8ca0339811476052038fc)
```
 "users": [
          "AndrewKushnir",
          "alxhub",
          "atscott",
          "mhevery",
          "pkozlowski-opensource",
          "~kara"
        ],
```

IgorMinar commented:
merge-assistance: I'm the code owner of the pullapprove config, and #36915 is merged we don't have a great way to approve this PR.

IgorMinar commented:
actually @josephperrott could give me a global approval.

josephperrott reviewed (approved):
LGTM

Reviewed-For: global-approvers

IgorMinar committed:
ci: special case tooling-cli-shared-api review group

The new tooling-cli-shared-api is used to guard changes to packages/compiler-cli/src/tooling.ts
which is a private API sharing channel between Angular FW and CLI.

Changes to this file should be rare and explicitly approved by at least two members
of the CLI team.

IgorMinar commented:
I've rebased this PR and added a new commit introducing the tooling-cli-shared-api group.

mary-poppins commented:
You can preview ef2fd5f at https://pr37467-ef2fd5f.ngbuilds.io/.
